Mathlib/Topology/MetricSpace/HolderNorm.lean

```diff
@@ -0,0 +1,228 @@
+
+lemma MemHolder.holderWith {r : ℝ≥0} {f : X → Y} (hf : MemHolder r f) :
+    HolderWith (nnHolderNorm r f) r f := by
+  intros x₁ x₂
+  by_cases hx : x₁ = x₂
+  · simp only [hx, edist_self, zero_le]
+  rw [nnHolderNorm, eHolderNorm, coe_toNNReal]
+  swap; exact hf.eHolderNorm_lt_top.ne
+  have h₁ : edist x₁ x₂ ^ (r : ℝ) ≠ 0 :=
+    (Ne.symm <| ne_of_lt <| ENNReal.rpow_pos (edist_pos.2 hx) (edist_lt_top x₁ x₂).ne)
+  have h₂ : edist x₁ x₂ ^ (r : ℝ) ≠ ∞ := by
+    simp [(edist_lt_top x₁ x₂).ne]
+  rw [← ENNReal.div_le_iff h₁ h₂]
+  refine le_iInf₂ fun C hC => ?_
+  rw [ENNReal.div_le_iff h₁ h₂]
+  exact hC x₁ x₂
+
+lemma memHolder_iff_holderWith {r : ℝ≥0} {f : X → Y} :
+    MemHolder r f ↔ HolderWith (nnHolderNorm r f) r f :=
+  ⟨MemHolder.holderWith, HolderWith.memHolder⟩
+
+lemma coe_nnHolderNorm_le_eHolderNorm {r : ℝ≥0} {f : X → Y} :
+    (nnHolderNorm r f : ℝ≥0∞) ≤ eHolderNorm r f :=
+  coe_toNNReal_le_self
+
+lemma HolderWith.coe_nnHolderNorm_eq_eHolderNorm
+    {C r : ℝ≥0} {f : X → Y} (hf : HolderWith C r f) :
+    (nnHolderNorm r f : ℝ≥0∞) = eHolderNorm r f := by
+  rw [nnHolderNorm, coe_toNNReal]
+  exact ne_of_lt <| lt_of_le_of_lt hf.eHolderNorm_le <| coe_lt_top (r := C)
+
+lemma MemHolder.coe_nnHolderNorm_eq_eHolderNorm
+    {r : ℝ≥0} {f : X → Y} (hf : MemHolder r f) :
+    (nnHolderNorm r f : ℝ≥0∞) = eHolderNorm r f :=
+  hf.holderWith.coe_nnHolderNorm_eq_eHolderNorm
+
+lemma HolderWith.nnholderNorm_le {C r : ℝ≥0} {f : X → Y} (hf : HolderWith C r f) :
+    nnHolderNorm r f ≤ C := by
+  rw [← ENNReal.coe_le_coe, hf.coe_nnHolderNorm_eq_eHolderNorm]
+  exact hf.eHolderNorm_le
+
+lemma MemHolder.comp {r : ℝ≥0} {Z : Type*} [MetricSpace Z] {f : Z → X} {g : X → Y}
+    (hf : MemHolder r f) (hg : MemHolder r g) : MemHolder (r * r) (g ∘ f) :=
+  (hg.holderWith.comp hf.holderWith).memHolder
+
+end MetricSpace
+
+section SeminormedAddCommGroup
+
+variable [MetricSpace X] [NormedAddCommGroup Y]
+variable {C r : ℝ≥0} {f g : X → Y}
+
+lemma MemHolder.add (hf : MemHolder r f) (hg : MemHolder r g) : MemHolder r (f + g) :=
+  (hf.holderWith.add hg.holderWith).memHolder
+
+lemma MemHolder.smul {α} [NormedDivisionRing α] [Module α Y] [BoundedSMul α Y]
+    (c : α) (hf : MemHolder r f) : MemHolder r (c • f) :=
+  (hf.holderWith.smul c).memHolder
+
+variable (r f g) in
+lemma eHolderNorm_add :
+    eHolderNorm r (f + g) ≤ eHolderNorm r f + eHolderNorm r g := by
+  by_cases hfg : MemHolder r f  ∧ MemHolder r g
+  · obtain ⟨hf, hg⟩ := hfg
+    rw [← hf.coe_nnHolderNorm_eq_eHolderNorm, ← hg.coe_nnHolderNorm_eq_eHolderNorm, ← coe_add]
+    exact (hf.add hg).holderWith.eHolderNorm_le.trans <|
+      coe_le_coe.2 (hf.holderWith.add hg.holderWith).nnholderNorm_le
+  · rw [Classical.not_and_iff_or_not_not, not_memHolder, not_memHolder] at hfg
+    obtain (h | h) := hfg
+    all_goals simp [h]
+
+lemma eHolderNorm_smul {α} [NormedDivisionRing α] [Module α Y] [BoundedSMul α Y] (c : α) :
+    eHolderNorm r (c • f) = ‖c‖₊ * eHolderNorm r f := by
```

Comment:
Can we also have the `nsmul` version assuming `BoundedSMul ℝ Y`?

Reply:
I'm not sure what you mean. Something like this? 
```lean
lemma eHoelderNorm_nsmul [Module ℝ Y] [BoundedSMul ℝ Y] (c : ℝ) :
    eHoelderNorm r (c • f) = ‖c‖₊ * eHoelderNorm r f
```

Reply:
No, something like
```
lemma eHoelderNorm_nsmul [Module ℝ Y] [BoundedSMul ℝ Y] (n : ℕ) :
    eHoelderNorm r (n • f) = n • eHoelderNorm r f
```